File: packages/core/src/wire/canonical-json.ts
/**
 * RFC 8785 JSON Canonicalization Scheme (JCS).
 * Produces a deterministic byte-form for wire-hash idempotency.
 *
 * Rules:
 * - Keys sorted lexicographically by UTF-16 code unit (per RFC 8785 §3.2.3).
 * - No whitespace.
 * - Integers stay integer; non-integer finite numbers use JS `String(n)` form.
 * - Rejects `NaN` / `Infinity` (throws TypeError pre-emission).
 * - Strings escape per RFC 8259 §7 with shortest escape sequences.
 * - Integer guard: throws TypeError if any integer is outside Number.MAX_SAFE_INTEGER ±.
 *   Callers wanting BigInt-style strings must pre-convert.
 */

/**
 * Encodes a single string value with RFC 8259 §7 minimal escaping.
 * Control characters < 0x20 are \uXXXX encoded.
 */
function encodeString(s: string): string {
  let result = '"';
  for (let i = 0; i < s.length; i++) {
    const c = s[i];
    const code = s.charCodeAt(i);
    if (c === '"') {
      result += '\\"';
    } else if (c === '\\') {
      result += '\\\\';
    } else if (c === '\b') {
      result += '\\b';
    } else if (c === '\f') {
      result += '\\f';
    } else if (c === '\n') {
      result += '\\n';
    } else if (c === '\r') {
      result += '\\r';
    } else if (c === '\t') {
      result += '\\t';
    } else if (code < 0x20) {
      result += '\\u' + code.toString(16).padStart(4, '0');
    } else {
      result += c;
    }
  }
  result += '"';
  return result;
}

/**
 * Core recursive serialiser implementing RFC 8785 JCS.
 *
 * @throws {TypeError} on NaN, Infinity, unsafe integers, or undefined values.
 */
function serializeValue(value: unknown): string {
  if (value === null) {
    return 'null';
  }

  if (value === undefined) {
    throw new TypeError('canonical-json: undefined is not a valid JSON value');
  }

  if (typeof value === 'boolean') {
    return value ? 'true' : 'false';
  }

  if (typeof value === 'number') {
    if (!isFinite(value)) {
      throw new TypeError(
        `canonical-json: non-finite number (${value}) cannot be serialized to JSON`,
      );
    }
    if (Number.isInteger(value)) {
      if (Math.abs(value) > Number.MAX_SAFE_INTEGER) {
        throw new TypeError(
          `canonical-json: integer ${value} is outside safe integer range; pre-convert to string`,
        );
      }
      return String(value);
    }
    return String(value);
  }

  if (typeof value === 'string') {
    return encodeString(value);
  }

  if (Array.isArray(value)) {
    if (value.length === 0) return '[]';
    const items = value.map((item) => serializeValue(item));
    return '[' + items.join(',') + ']';
  }

  if (typeof value === 'object') {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).sort();
    if (keys.length === 0) return '{}';
    const pairs = keys
      .filter((k) => obj[k] !== undefined)
      .map((k) => encodeString(k) + ':' + serializeValue(obj[k]));
    return '{' + pairs.join(',') + '}';
  }

  throw new TypeError(`canonical-json: unsupported value type: ${typeof value}`);
}

/**
 * Produce a deterministic RFC 8785 JSON string.
 *
 * @throws {TypeError} on NaN, Infinity, unsafe integers, or undefined.
 */
export function canonicalize(value: unknown): string {
  return serializeValue(value);
}

/**
 * Convenience: parse-then-canonicalize, for normalizing arbitrary JSON input.
 *
 * @throws {SyntaxError} if `raw` is not valid JSON.
 * @throws {TypeError} if the parsed value contains non-finite numbers or unsafe integers.
 */
export function canonicalizeRaw(raw: string): string {
  return canonicalize(JSON.parse(raw) as unknown);
}

File: packages/adapter-python-process/src/pkit_wire/canonical_json.py
"""canonical_json — RFC 8785 JSON Canonicalization Scheme (JCS).

Produces deterministic UTF-8 bytes for wire-hash idempotency:
  - Keys sorted lexicographically (sort_keys=True)
  - No whitespace (separators=(',',':'))
  - ensure_ascii=False preserves non-ASCII UTF-8 codepoints
  - Integers outside [-MAX_SAFE_INTEGER, MAX_SAFE_INTEGER] are rejected

MAX_SAFE_INTEGER = 9_007_199_254_740_991 (2^53 - 1 per ECMAScript)
Callers wanting BigInt-style integers MUST pre-convert to string.

Reference: packages/core/src/wire/canonical-json.ts
"""

from __future__ import annotations

import json
from typing import Any

MAX_SAFE_INTEGER = 9_007_199_254_740_991


def _check_integers(obj: Any) -> None:
    """Recursively validate all integers in obj are within MAX_SAFE_INTEGER."""
    if isinstance(obj, bool):
        # bool is subclass of int; booleans are wire-safe
        return
    if isinstance(obj, int):
        if abs(obj) > MAX_SAFE_INTEGER:
            raise TypeError(
                f"canonical_json: integer {obj} is outside safe integer range "
                f"[-{MAX_SAFE_INTEGER}, {MAX_SAFE_INTEGER}]; pre-convert to string"
            )
    elif isinstance(obj, dict):
        for v in obj.values():
            _check_integers(v)
    elif isinstance(obj, list):
        for item in obj:
            _check_integers(item)


def canonical_json(obj: Any) -> bytes:
    """Encode obj to RFC 8785 canonical JSON bytes.

    Args:
        obj: Any JSON-serialisable Python value.

    Returns:
        UTF-8 encoded canonical JSON bytes with sorted keys and no whitespace.

    Raises:
        TypeError: if any integer exceeds MAX_SAFE_INTEGER, or obj contains
                   non-JSON-serialisable types (NaN, Infinity, etc.).
    """
    _check_integers(obj)
    return json.dumps(
        obj,
        separators=(",", ":"),
        sort_keys=True,
        ensure_ascii=False,
    ).encode("utf-8")
